File: src/06.hashMap/07.290.go
package hashmap

import "strings"

func WordPattern(pattern string, s string) bool {
	words := strings.Split(s, " ")

	if len(words) != len(pattern) {
		return false
	}

	w2p := map[string]byte{}
	p2w := map[byte]string{}

	for i, w := range words {
		ch := pattern[i]

		if v, ok := w2p[w]; ok && v != ch {
			return false
		}

		if v, ok := p2w[ch]; ok && v != w {
			return false
		}

		w2p[w] = ch
		p2w[ch] = w
	}

	return true
}

File: src/06.hashMap/07.290.ts
export function wordPattern(pattern: string, s: string): boolean {
    const words = s.split(" ");

    if (words.length !== pattern.length) {
        return false;
    }

    const w2p = new Map<string, string>();
    const p2w = new Map<string, string>();

    for (let i = 0; i < words.length; i++) {
        const word = words[i];
        const ch = pattern[i];
        if (
            (w2p.has(word) && w2p.get(word) !== ch) ||
            (p2w.has(ch) && p2w.get(ch) !== word)
        ) {
            return false;
        }

        w2p.set(word, ch);
        p2w.set(ch, word);
    }

    return true;
}
